keras/src/callbacks/orbax_checkpoint.py

```diff
@@ -0,0 +1,702 @@
+# ocp = LazyModule("orbax.checkpoint"); ocp.CheckpointManager
+
+
+def _get_state_tree(model):
+    """Get the complete model state as a nested tree structure."""
+    state_tree = model.get_state_tree(value_format="numpy_array")
+
+    # Convert numpy scalar types to Python types for Orbax compatibility
+    def convert_scalars(obj):
+        if isinstance(obj, np.ndarray) and obj.ndim == 0:
+            # Convert 0-dimensional numpy arrays (scalars) to Python types
+            return obj.item()
+        elif isinstance(obj, np.generic):
+            # Convert numpy scalar types (like np.float32) to Python types
+            return obj.item()
+        elif isinstance(obj, dict):
+            return {k: convert_scalars(v) for k, v in obj.items()}
+        else:
+            return obj
+
+    return convert_scalars(state_tree)
+
+
+def _flatten_state_tree_values(state_tree):
+    """Flatten nested state tree into a list of values in consistent order."""
+    values = []
+
+    def _flatten(obj):
+        if isinstance(obj, dict):
+            for key in sorted(obj.keys()):  # Sort for consistent ordering
+                _flatten(obj[key])
+        else:
+            # Save any non-dict value (numpy arrays, lists, scalars, etc.)
+            values.append(obj)
+
+    _flatten(state_tree)
+    return values
+
+
+def _reconstruct_state_tree_with_values(structure, values):
+    """Reconstruct state tree structure with provided values."""
+    value_iter = iter(values)
+
+    def _reconstruct(obj):
+        if isinstance(obj, dict):
+            new_dict = {}
+            for key in sorted(obj.keys()):
+                new_dict[key] = _reconstruct(obj[key])
+            return new_dict
+        else:
+            value = next(value_iter)
+            # Handle different cases for value conversion
+            if isinstance(obj, np.generic):
+                # obj is a numpy scalar (0-dimensional)
+                if isinstance(value, (int, float)):
+                    # Convert Python scalar to numpy scalar
+                    return np.array(value, dtype=obj.dtype)
+                elif isinstance(value, np.ndarray):
+                    # value is a numpy array, convert to scalar if needed
+                    if value.ndim == 0:
+                        return np.array(value.item(), dtype=obj.dtype)
+                    elif value.ndim == 1 and value.size == 1:
+                        return np.array(value.item(), dtype=obj.dtype)
+                    else:
+                        return value.astype(obj.dtype).reshape(obj.shape)
+                else:
+                    return np.array(value, dtype=obj.dtype)
+            elif isinstance(obj, np.ndarray):
+                # obj is a numpy array
+                if isinstance(value, np.ndarray):
+                    return value.astype(obj.dtype).reshape(obj.shape)
+                else:
+                    return np.array(value, dtype=obj.dtype).reshape(obj.shape)
```

Comment:
I'm confused why you need these. And doesn't Orbax give you JAX arrays? What about the sharding?

Reply:
You can and should pass Jax arrays with the Jax backend, that's how the sharding is supported: https://orbax.readthedocs.io/en/latest/guides/checkpoint/checkpointing_pytrees.html#basic-checkpointing